### PR TITLE
Allow to set a CA certificate for LDAPS connections

### PR DESCRIPTION
**Pull request by rhafer**
This introduces the new environment variable LDAP_TLS_CACERT, which
allows to configure a CA certificate file for verifying the LDAP
server's TLS certificate.

Also the ServerName is now set properly in the TLSConfig, to allow the
verification to succeed.

Closes: #28

### CODE REVIEW

**Review by longsleep — CHANGES_REQUESTED** [review]
@rhafer i am not sure why this is required. Please check if the functionality already provided by Go is not sufficient for the use case which this is supposed to solve.

```
// On Unix systems other than macOS the environment variables SSL_CERT_FILE and
// SSL_CERT_DIR can be used to override the system default locations for the SSL
// certificate file and SSL certificate files directory, respectively. The
// latter can be a colon-separated list.
```

**Comment by rhafer** [comment]
> i am not sure why this is required. Please check if the functionality already provided by Go is not sufficient for the use case which this is supposed to solve.

@longsleep Thanks for the pointer. I actually overlooked `SSL_CERT_FILE`. Though there is at least some desire to make our code work on darwin and it seem SSL_CERT_FILE does not work there.
Also I would prefer to be explicit here an just override the cert for the LDAP client. Using `SSL_CERT_FILE` might have unwanted effects as it would override the settings for all tls clients that lico might be using.

File: utils/http.go [hunk 1]
```diff
@@ -82,6 +86,22 @@ func InsecureSkipVerifyTLSConfig() *tls.Config {
 	return config
 }
 
+// TlsConfigCACert returns a new tls.Config with caCertFile added to the set of trusted CAs
+func TlsConfigCACert(caCertFile string) (*tls.Config, error) {
```
Comment (longsleep) [inline utils/http.go:90]: Should be uppercase TLS

File: utils/http.go [hunk 2]
```diff
@@ -19,13 +19,17 @@ package utils
...
+	"fmt"
+	"io/ioutil"
 	"net"
 	"net/http"
 	"time"
 
 	"golang.org/x/net/http2"
 
 	"github.com/libregraph/lico/version"
+	"github.com/pkg/errors"
```
Comment (longsleep) [inline utils/http.go:32]: Wrong import group, please move up to the external group.

File: bootstrap/backends/ldap/ldap.go [hunk 3]
```diff
@@ -87,9 +90,19 @@ func NewIdentityManager(bs bootstrap.Bootstrap) (identity.Manager, error) {
 		subMapping = strings.Split(subMappingString, " ")
 	}
 
+	var tlsConfig *tls.Config
+	if caCertFile := os.Getenv("LDAP_TLS_CACERT"); caCertFile != "" {
+		var err error
+		if tlsConfig, err = utils.TlsConfigCACert(caCertFile); err != nil {
```
Comment (longsleep) [inline bootstrap/backends/ldap/ldap.go:96]: This is not using the TLSConfig from config which might be confusing.

File: identifier/backends/ldap/ldap.go [hunk 4]
```diff
@@ -305,6 +305,7 @@ func NewLDAPIdentifierBackend(
 		if uri.Port() == "" {
 			addr += ":636"
 		}
```
Comment (longsleep) [inline identifier/backends/ldap/ldap.go:307]: This change seems unrelated to the mr. At least it should be a dedicated commit.

File: utils/http.go [hunk 5]
```diff
@@ -82,6 +86,22 @@ func InsecureSkipVerifyTLSConfig() *tls.Config {
...
 
+// TlsConfigCACert returns a new tls.Config with caCertFile added to the set of trusted CAs
+func TlsConfigCACert(caCertFile string) (*tls.Config, error) {
+	config := DefaultTLSConfig()
+	if pemBytes, err := ioutil.ReadFile(caCertFile); err == nil {
+		rpool, _ := x509.SystemCertPool()
+		if rpool.AppendCertsFromPEM(pemBytes) {
+			config.RootCAs = rpool
+		} else {
+			return nil, fmt.Errorf("Error appending CA certificate '%s' to pool", caCertFile)
```
Comment (longsleep) [inline utils/http.go:97]: This can add more than one certificate - so better use plural in error message - here and.below.

**Review by longsleep — CHANGES_REQUESTED** [review]

**Comment by longsleep** [comment]
> > i am not sure why this is required. Please check if the functionality already provided by Go is not sufficient for the use case which this is supposed to solve.
> 
> @longsleep Thanks for the pointer. I actually overlooked `SSL_CERT_FILE`. Though there is at least some desire to make our code work on darwin and it seem SSL_CERT_FILE does not work there. Also I would prefer to be explicit here an just override the cert for the LDAP client. Using `SSL_CERT_FILE` might have unwanted effects as it would override the settings for all tls clients that lico might be using.

I think its less confusing to have a "for everything"

File: utils/http.go (hunk 2 above)
Comment (rhafer) [reply utils/http.go:32]: Hm, I'll blame the vim-go plugin here :-) will adjust that.

File: bootstrap/backends/ldap/ldap.go (hunk 3 above)
Comment (rhafer) [reply bootstrap/backends/ldap/ldap.go]: I've reworked that. It well now alway work on a clone of  TLSConfig from config

File: identifier/backends/ldap/ldap.go (hunk 4 above)
Comment (rhafer) [reply identifier/backends/ldap/ldap.go:307]: Done

File: bootstrap/backends/ldap/ldap.go [hunk 6]
```diff
@@ -87,9 +91,24 @@ func NewIdentityManager(bs bootstrap.Bootstrap) (identity.Manager, error) {
...
 
+	// Use a clone here to avoid changing the config of other possible users of the config.
+	tlsConfig := config.TLSClientConfig.Clone()
+	if caCertFile := os.Getenv("LDAP_TLS_CACERT"); caCertFile != "" {
+		if pemBytes, err := ioutil.ReadFile(caCertFile); err == nil {
+			rpool, _ := x509.SystemCertPool()
+			if rpool.AppendCertsFromPEM(pemBytes) {
+				tlsConfig.RootCAs = rpool
+			} else {
+				return nil, fmt.Errorf("Error appending CA certificate(s) '%s' to pool", caCertFile)
```
Comment (longsleep) [inline bootstrap/backends/ldap/ldap.go:102]: error strings should start lower case, to be in line with the other error messages in lico, use something like `failed to append CA certificates from '%s' to tool`.

Comment (longsleep) [inline bootstrap/backends/ldap/ldap.go:102]: error strings should start lower case, to be in line with the other error messages in lico, use something like `failed to append CA certificates from '%s' to tool`.

File: bootstrap/backends/ldap/ldap.go [hunk 7]
```diff
@@ -18,10 +18,14 @@
 package bsldap
 
 import (
+	"crypto/x509"
 	"fmt"
+	"io/ioutil"
 	"os"
 	"strings"
 
+	"github.com/pkg/errors"
```
Comment (longsleep) [inline bootstrap/backends/ldap/ldap.go:27]: Assuming we really want this package, an update to go.mod is missing. Please run `go mod tidy`.

Comment (longsleep) [inline bootstrap/backends/ldap/ldap.go:27]: Assuming we really want this package, an update to go.mod is missing. Please run `go mod tidy`.

File: bootstrap/backends/ldap/ldap.go [hunk 8]
```diff
@@ -87,9 +91,24 @@ func NewIdentityManager(bs bootstrap.Bootstrap) (identity.Manager, error) {
...
+	if caCertFile := os.Getenv("LDAP_TLS_CACERT"); caCertFile != "" {
+		if pemBytes, err := ioutil.ReadFile(caCertFile); err == nil {
+			rpool, _ := x509.SystemCertPool()
+			if rpool.AppendCertsFromPEM(pemBytes) {
+				tlsConfig.RootCAs = rpool
+			} else {
+				return nil, fmt.Errorf("Error appending CA certificate(s) '%s' to pool", caCertFile)
+			}
+		} else {
+			return nil, errors.Wrapf(err, "Error reading CA certificate '%s'", caCertFile)
```
Comment (longsleep) [inline bootstrap/backends/ldap/ldap.go:105]: no need to use the `errors` package. Instead use error wrapping via the [fmt package](https://pkg.go.dev/fmt#Errorf) (since Go 1.13, lico requires Go 1.16).  Also please rework the error messages (lowercase, similar style to the existing messages).

Comment (longsleep) [inline bootstrap/backends/ldap/ldap.go:105]: no need to use the `errors` package. Instead use error wrapping via the [fmt package](https://pkg.go.dev/fmt#Errorf) (since Go 1.13, lico requires Go 1.16).

**Review by longsleep — CHANGES_REQUESTED** [review]

**Review by longsleep — CHANGES_REQUESTED** [review]

File: bootstrap/backends/ldap/ldap.go [hunk 9]
```diff
@@ -87,9 +89,24 @@ func NewIdentityManager(bs bootstrap.Bootstrap) (identity.Manager, error) {
...
 
+	// Use a clone here to avoid changing the config of other possible users of the config.
+	tlsConfig := config.TLSClientConfig.Clone()
+	if caCertFile := os.Getenv("LDAP_TLS_CACERT"); caCertFile != "" {
+		if pemBytes, err := ioutil.ReadFile(caCertFile); err == nil {
+			rpool, _ := x509.SystemCertPool()
+			if rpool.AppendCertsFromPEM(pemBytes) {
+				tlsConfig.RootCAs = rpool
+			} else {
+				return nil, fmt.Errorf("faled to append CA certificate(s) from '%s' to pool", caCertFile)
```
Comment (longsleep) [inline bootstrap/backends/ldap/ldap.go:100]: Typo faled

**Review by longsleep — CHANGES_REQUESTED** [review]

**Review by longsleep — APPROVED** [review]
